File: app/src/main/java/com/veganbeauty/app/features/community/com_feed/CommunityDiscoverPeopleFragment.java
package com.veganbeauty.app.features.community.com_feed;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;

import com.veganbeauty.app.core.base.RootieFragment;
import com.veganbeauty.app.data.local.LocalJsonReader;
import com.veganbeauty.app.data.local.RootieDatabase;
import com.veganbeauty.app.data.local.dao.CommunityDao;
import com.veganbeauty.app.data.local.entities.UserEntity;
import com.veganbeauty.app.data.local.entities.UserMemoryEntity;
import com.veganbeauty.app.data.remote.FirestoreService;
import com.veganbeauty.app.data.repository.CommunityRepository;
import com.veganbeauty.app.databinding.ComFragmentDiscoverPeopleBinding;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class CommunityDiscoverPeopleFragment extends RootieFragment {

    private ComFragmentDiscoverPeopleBinding binding;
    private CommunityViewModel viewModel;
    private CommunityDao communityDao;
    private final FirestoreService firestoreService = new FirestoreService();
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    private void handleUserAction(UserEntity user, String action) {
        String content = "FOLLOW".equals(action) ? "Bạn đã bắt đầu theo dõi " + user.getUsername() : "Bạn đã chấp nhận yêu cầu theo dõi của " + user.getUsername();
        UserMemoryEntity memory = new UserMemoryEntity(
                java.util.UUID.randomUUID().toString(),
                action,
                user.getUser_id(),
                user.getUsername(),
                user.getAvatar() != null ? user.getAvatar() : "",
                content,
                System.currentTimeMillis()
        );
        executor.execute(() -> {
            try {
                communityDao.insertUserMemory(memory);
                firestoreService.uploadUserMemory(memory);
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }

    private final DiscoverPeopleAdapter suggestAdapter = new DiscoverPeopleAdapter(new ArrayList<>(), DiscoverPeopleAdapter.ActionType.SUGGEST, this::handleUserAction);
    private final DiscoverPeopleAdapter requestAdapter = new DiscoverPeopleAdapter(new ArrayList<>(), DiscoverPeopleAdapter.ActionType.REQUEST, this::handleUserAction);
    private final DiscoverPeopleAdapter followBackAdapter = new DiscoverPeopleAdapter(new ArrayList<>(), DiscoverPeopleAdapter.ActionType.FOLLOW_BACK, this::handleUserAction);

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = ComFragmentDiscoverPeopleBinding.inflate(inflater, container, false);
        setupViewModel();
        return binding.getRoot();
    }

    private void setupViewModel() {
        RootieDatabase db = RootieDatabase.getDatabase(requireContext());
        communityDao = db.communityDao();
        CommunityRepository repository = new CommunityRepository(communityDao, new LocalJsonReader(requireContext()), new FirestoreService());
        CommunityViewModelFactory factory = new CommunityViewModelFactory(repository);
        viewModel = new ViewModelProvider(this, factory).get(CommunityViewModel.class);
    }

    @Override
    public void setupUI(@NonNull View view) {
        binding.ivBack.setOnClickListener(v -> getParentFragmentManager().popBackStack());

        binding.rvSuggestions.setAdapter(suggestAdapter);
        binding.rvRequests.setAdapter(requestAdapter);
        binding.rvFollowBacks.setAdapter(followBackAdapter);
    }

    @Override
    protected void observeViewModel() {
        viewModel.getUsers().observe(getViewLifecycleOwner(), users -> {
            if (users != null && !users.isEmpty()) {
                executor.execute(() -> {
                    try {
                        LocalJsonReader jsonReader = new LocalJsonReader(requireContext());
                        Map<String, List<String>> mySocialData = jsonReader.getSocialDataForUser("test_001");
                        
                        Set<String> myFriends = new HashSet<>();
                        if (mySocialData.containsKey("friends")) myFriends.addAll(mySocialData.get("friends"));
                        
                        List<String> suggestIds = mySocialData.containsKey("suggested") ? mySocialData.get("suggested") : new ArrayList<>();
                        List<String> requestIds = mySocialData.containsKey("friend_requests") ? mySocialData.get("friend_requests") : new ArrayList<>();
                        
                        Set<String> followers = new HashSet<>();
                        if (mySocialData.containsKey("followers")) followers.addAll(mySocialData.get("followers"));
                        
                        Set<String> following = new HashSet<>();
                        if (mySocialData.containsKey("following")) following.addAll(mySocialData.get("following"));
                        
                        Set<String> followBackIds = new HashSet<>(followers);
                        followBackIds.removeAll(following);
                        followBackIds.removeAll(myFriends);
                        
                        List<UserEntity> suggestions = new ArrayList<>();
                        List<UserEntity> requests = new ArrayList<>();
                        List<UserEntity> followBacks = new ArrayList<>();
                        
                        for (UserEntity user : users) {
                            if (!"test_001".equals(user.getUser_id())) {
                                if (suggestIds.contains(user.getUser_id())) suggestions.add(user);
                                if (requestIds.contains(user.getUser_id())) requests.add(user);
                                if (followBackIds.contains(user.getUser_id())) followBacks.add(user);
                            }
                        }
                        
                        List<String> allTargetIds = new ArrayList<>();
                        for (UserEntity u : suggestions) allTargetIds.add(u.getUser_id());
                        for (UserEntity u : requests) allTargetIds.add(u.getUser_id());
                        for (UserEntity u : followBacks) allTargetIds.add(u.getUser_id());
                        
                        Map<String, List<String>> mutualMap = jsonReader.getMutualFriendsForUsers(new ArrayList<>(myFriends), allTargetIds);
                        
                        applyMutualData(suggestions, mutualMap, users);
                        applyMutualData(requests, mutualMap, users);
                        applyMutualData(followBacks, mutualMap, users);
                        
                        requireActivity().runOnUiThread(() -> {
                            if (isAdded()) {
                                suggestAdapter.updateData(suggestions);
                                requestAdapter.updateData(requests);
                                followBackAdapter.updateData(followBacks);
                            }
                        });
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                });
            }
        });
    }
    
    private void applyMutualData(List<UserEntity> list, Map<String, List<String>> mutualMap, List<UserEntity> allUsers) {
        for (UserEntity user : list) {
            List<String> mutualIds = mutualMap.get(user.getUser_id());
            if (mutualIds == null) mutualIds = new ArrayList<>();
            
            user.setMutualCount(mutualIds.size());
            if (!mutualIds.isEmpty()) {
                List<UserEntity> mutualUsers = new ArrayList<>();
                for (UserEntity u : allUsers) {
                    if (mutualIds.contains(u.getUser_id())) {
                        mutualUsers.add(u);
                    }
                }
                
                if (!mutualUsers.isEmpty()) {
                    String name = mutualUsers.get(0).getFull_name();
                    if (name == null || name.trim().isEmpty()) {
                        name = mutualUsers.get(0).getUsername();
                    }
                    user.setFirstMutualFriendName(name);
                    
                    List<String> avatars = new ArrayList<>();
                    for (int i = 0; i < Math.min(3, mutualUsers.size()); i++) {
                        String avatar = mutualUsers.get(i).getAvatar();
                        if (avatar != null) avatars.add(avatar);
                    }
                    user.setMutualFriendAvatars(avatars);
                }
            }
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/veganbeauty/app/features/community/com_feed/CommunityDiscoverPeopleFragment.kt
package com.veganbeauty.app.features.community.com_feed

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import androidx.room.Room
import com.veganbeauty.app.core.base.RootieFragment
import com.veganbeauty.app.data.local.LocalJsonReader
import com.veganbeauty.app.data.local.RootieDatabase
import com.veganbeauty.app.data.remote.FirestoreService
import com.veganbeauty.app.data.repository.CommunityRepository
import com.veganbeauty.app.databinding.ComFragmentDiscoverPeopleBinding
import com.veganbeauty.app.R
import kotlinx.coroutines.launch

class CommunityDiscoverPeopleFragment : RootieFragment() {

    private var _binding: ComFragmentDiscoverPeopleBinding? = null
    private val binding get() = _binding!!

    private lateinit var viewModel: CommunityViewModel
    private lateinit var communityDao: com.veganbeauty.app.data.local.dao.CommunityDao
    private val firestoreService = FirestoreService()

    private fun handleUserAction(user: com.veganbeauty.app.data.local.entities.UserEntity, action: String) {
        val content = if (action == "FOLLOW") "Bạn đã bắt đầu theo dõi ${user.username}" else "Bạn đã chấp nhận yêu cầu theo dõi của ${user.username}"
        val memory = com.veganbeauty.app.data.local.entities.UserMemoryEntity(
            actionType = action,
            targetUserId = user.user_id,
            targetUsername = user.username,
            targetAvatar = user.avatar ?: "",
            content = content
        )
        kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO).launch {
            // Save to local SQLite
            communityDao.insertUserMemory(memory)
            // Upload to Firebase
            firestoreService.uploadUserMemory(memory)
        }
    }

    private val suggestAdapter = DiscoverPeopleAdapter(emptyList(), DiscoverPeopleAdapter.ActionType.SUGGEST) { user, act -> handleUserAction(user, act) }
    private val requestAdapter = DiscoverPeopleAdapter(emptyList(), DiscoverPeopleAdapter.ActionType.REQUEST) { user, act -> handleUserAction(user, act) }
    private val followBackAdapter = DiscoverPeopleAdapter(emptyList(), DiscoverPeopleAdapter.ActionType.FOLLOW_BACK) { user, act -> handleUserAction(user, act) }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = ComFragmentDiscoverPeopleBinding.inflate(inflater, container, false)
        setupViewModel()
        return binding.root
    }

    private fun setupViewModel() {
        val db = RootieDatabase.getDatabase(requireContext())
        communityDao = db.communityDao()
        val repository = CommunityRepository(communityDao, LocalJsonReader(requireContext()), FirestoreService())
        val factory = CommunityViewModelFactory(repository)
        viewModel = ViewModelProvider(this, factory)[CommunityViewModel::class.java]
    }

    override fun setupUI(view: View) {
        binding.ivBack.setOnClickListener {
            parentFragmentManager.popBackStack()
        }

        binding.rvSuggestions.adapter = suggestAdapter
        binding.rvRequests.adapter = requestAdapter
        binding.rvFollowBacks.adapter = followBackAdapter
    }

    override fun observeViewModel() {
        viewModel.users.observe(viewLifecycleOwner) { users ->
            if (users.isNotEmpty()) {
                val jsonReader = com.veganbeauty.app.data.local.LocalJsonReader(requireContext())
                val mySocialData = jsonReader.getSocialDataForUser("test_001")
                val myFriends = (mySocialData["friends"] ?: emptyList()).toSet()
                
                val suggestIds = mySocialData["suggested"] ?: emptyList()
                val requestIds = mySocialData["friend_requests"] ?: emptyList()
                // For "Follow Back" (Theo dõi lại), these are people in "followers" but NOT in "following" and NOT "friends"
                val followers = (mySocialData["followers"] ?: emptyList()).toSet()
                val following = (mySocialData["following"] ?: emptyList()).toSet()
                val followBackIds = followers.subtract(following).subtract(myFriends).toList()
                
                val suggestions = users.filter { it.user_id in suggestIds && it.user_id != "test_001" }
                val requests = users.filter { it.user_id in requestIds && it.user_id != "test_001" }
                val followBacks = users.filter { it.user_id in followBackIds && it.user_id != "test_001" }
                
                // Calculate mutual friends
                val allTargetIds = suggestions.map { it.user_id } + requests.map { it.user_id } + followBacks.map { it.user_id }
                val mutualMap = jsonReader.getMutualFriendsForUsers(myFriends, allTargetIds)
                
                val applyMutualData = { list: List<com.veganbeauty.app.data.local.entities.UserEntity> ->
                    list.forEach { user ->
                        val mutualIds = mutualMap[user.user_id] ?: emptyList()
                        user.mutualCount = mutualIds.size
                        if (mutualIds.isNotEmpty()) {
                            val mutualUsers = users.filter { it.user_id in mutualIds }
                            if (mutualUsers.isNotEmpty()) {
                                user.firstMutualFriendName = mutualUsers[0].full_name.ifBlank { mutualUsers[0].username }
                                user.mutualFriendAvatars = mutualUsers.mapNotNull { it.avatar }.take(3)
                            }
                        }
                    }
                }
                
                applyMutualData(suggestions)
                applyMutualData(requests)
                applyMutualData(followBacks)
                
                suggestAdapter.updateData(suggestions)
                requestAdapter.updateData(requests)
                followBackAdapter.updateData(followBacks)
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}

